Sources/SwiftProtobuf/SetPathDecoder.swift

```diff
@@ -0,0 +1,364 @@
+// Sources/SwiftProtobuf/SetPathDecoder.swift - Path decoder (Setter)
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/main/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+///
+/// Decoder which sets value of a field by its path.
+///
+// -----------------------------------------------------------------------------
+
+import Foundation
+
+/// Describes errors can occure during decoding a proto by path.
+public enum PathDecodingError: Error {
+
+  /// Describes a mismatch in type of the fields.
+  ///
+  /// If a value of type A is applied to a path with type B.
+  /// this error will be thrown.
+  case typeMismatch
+
+  /// Describes path is not found in message type.
+  ///
+  /// If a message has no field with the given path this
+  /// error will be thrown.
+  case pathNotFound
+}
+
+extension Message {
+  static func number(for field: String) -> Int? {
+    guard let type = Self.self as? _ProtoNameProviding.Type else {
+      return nil
+    }
+    return type._protobuf_nameMap.number(forJSONName: field)
+  }
+}
+
+struct SetPathDecoder<T: Message>: Decoder {
+
+  private let value: Any?
+  private var number: Int?
+  private let nextPath: [String]
+  private let replaceRepeatedFields: Bool
+
+  init(
+    path: [String],
+    value: Any?,
+    replaceRepeatedFields: Bool
+  ) {
+    if let firstComponent = path.first {
```

Comment:
Does this need to do any validation on the path?  If not, maybe add a comment about why it can assume the paths are always valid?

Reply:
I changed the code in a more understandable way, after checking for first path component, we check whether this component refers to a valid field number of message or not. If it doesn't, the number property remains nil and decoder does nothing. (As it has no `nextFieldNumber` to pass)